### PR TITLE
Add Rails 7 support

### PR DESCRIPTION
**Pull request by peterberkenbosch**
**Description**

With the release of Rails 7.0.0 we should start preparing Solidus edge to be supported. This PR starts with that.

Default using the latest released rails 7.0 on setup and sandbox creation:

```
bin/setup
bin/sandbox
```

There is however still an error showing up with title: **TypeError in Spree::Home#index**

```
no implicit conversion of String into Integer
```

Solving this with:

add `mini_magick` to `sandbox/config/application.rb` since the default `:vips` does not work out of the box with Solidus just yet. (see more details below).

```ruby
config.active_storage.variant_processor = :mini_magick
```

If we want to use `:vips` by default we have to change `core/app/models/concerns/spree/active_storage_adapter/attachment.rb#variant`: 

from:

```ruby
def variant(style = nil)
  size = style_to_size(style)
  @attachment.variant(
    resize_to_limit: size,
    strip: true
  ).processed
end
```

to:

```ruby
def variant(style = nil)
  size = style_to_size(style)
  @attachment.variant(
    resize_to_limit: size.map(&:to_i)
  ).processed
end
```

The size elements have to be an integer, and `strip` is not a valid command for `:vips`

Dependencies that we wait for new release:

- [x] ransack
- [x] devise
- [x] awesome_nested_set [tracking PR](https://github.com/collectiveidea/awesome_nested_set/issues/462)
- [x] font-awesome-rails
- [x] rails-i18n
- [x] canonical-rails support for Rails 7.0.1 [tracking PR](https://github.com/jumph4x/canonical-rails/pull/76)
- [x] `config.active_storage.variant_processor = :mini_magick` automation with install generator?
- [x] Rails ~> 7.0.0 depends on Ruby >= 2.7.0
- [x] Netlify & CircleCI update to Ruby >= 2.7.0
      

**Checklist:**
- [ ] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [ ] I have added a detailed description into each commit message
- [ ] I have updated Guides and README accordingly to this change (if needed)
- [ ] I have added tests to cover this change (if needed)
- [ ] I have attached screenshots to this PR for visual changes (if needed)

### CODE REVIEW

**Review by jarednorman — COMMENTED** [review]
Thanks for getting a head start on this @peterberkenbosch!

**Comment by peterberkenbosch** [comment]
Updated this with the last few missing dependencies to run the sandbox.

**Comment by peterberkenbosch** [comment]
Updated with the latest released gems. Wondering if we want to stop supporting Rails 5.2? Also not sure why the added circlci build for postgres_rails70 is not being picked up by circleCI, probably needs someone with more permissions on the CircleCI account to enable or approve that build step?

**Comment by jarednorman** [comment]
Oooh, look at that green check mark.

**Comment by BenMorganIO** [comment]
I know I'm not supposed to leave emoji specific Github comments, but

😎

**Comment by denys-medynskyi** [comment]
hey 👋 

Looking forward to Rails 7 🎉 

Anything I can do to help?
Do we have an ETA of merging this PR?

File: Gemfile
Comment (bbean86) [inline Gemfile:15]: Does this need to be adjusted to compare against any version later than 7.0.0 since 7.0.1 is already out?

Thank you for opening this PR, by the way, we'd love to bump to Rails 7 ASAP and Solidus is one of the few things preventing that for us.

**Comment by jarednorman** [comment]
@peterberkenbosch Now that specs are green, should we be taking a more serious look at this PR?

**Comment by peterberkenbosch** [comment]
Yes! Let's do that!

File: Gemfile
Comment (peterberkenbosch) [reply Gemfile:15]: Yes, this needs to be a bit smarter. Good catch!

File: core/lib/spree/core/controller_helpers/current_host.rb
Comment (jarednorman) [inline core/lib/spree/core/controller_helpers/current_host.rb:10]: Can you explain why you're doing this?

**Review by jarednorman — CHANGES_REQUESTED** [review]
Can you rebase out some of the Gemfile churn as the different versions of things get updated/released? We also don't need stuff like changing the version of devise and then removing it altogether both in the git history.

Left a couple small comments, but generally this all looks good. Thanks for taking this on, Peter. 🙏🏻

File: core/lib/spree/core/controller_helpers/current_host.rb
Comment (peterberkenbosch) [reply core/lib/spree/core/controller_helpers/current_host.rb:10]: The remove deprecation warning on Rails 7. Might be better suitable for standalone PR now that I think of it. So we have 1 PR focussed only on adding Rails 7 compatibility.

**Comment by waiting-for-dev** [comment]
> Wondering if we want to stop supporting Rails 5.2?

I would say YES! Thoughts? cc @kennyadsl

> not sure why the added circlci build for postgres_rails70 is not being picked up by circleCI

@peterberkenbosch, besides defining it, you need to add a new job within [`workflows -> build -> jobs`](https://github.com/solidusio/solidus/blob/1ab4826e1d0eb5687068ffe910c206efa5d7f717/.circleci/config.yml#L144).

However, what we need to add is a job for `postgres_rails61`, and update the default Rails version to `7.0` in the [Gemfile](https://github.com/solidusio/solidus/blob/1ab4826e1d0eb5687068ffe910c206efa5d7f717/Gemfile#L12).

File: core/lib/spree/core/controller_helpers/current_host.rb
Comment (waiting-for-dev) [reply core/lib/spree/core/controller_helpers/current_host.rb:10]: I think it's pertinent to have it on this PR. Besides, if there's a warning, CI will fail.

Comment (peterberkenbosch) [reply core/lib/spree/core/controller_helpers/current_host.rb:10]: ah.. thanks! Let's keep it then.

**Comment by kennyadsl** [comment]
For our current policy, we should stop supporting a specific Rails version when it no more receives security patches, and 5.2 appears to be still receiving them, see: https://guides.rubyonrails.org/maintenance_policy.html#severe-security-issues. Also, changing this will probably require a major bump for Solidus, because this would be technically a breaking change, isn't it?

File: Gemfile
Comment (kennyadsl) [inline Gemfile:16]: I have some concerns here. Apps that have `solidus_core` as a dependency will still be able to only get the RubyGems version of this gem (unless they add this line in their `Gemfile` explicitly) because when bundling, Bundler will only check the dependencies defined in the solidus's `gemspec`, and not this Gemfile.

File: bin/sandbox
Comment (kennyadsl) [inline bin/sandbox:56]: Now that https://github.com/collectiveidea/awesome_nested_set/pull/460 has been merged, why not using:

```
gem 'awesome_nested_set', github: 'collectiveidea/awesome_nested_set'
```

as you did in the Gemfile?

File: bin/sandbox
Comment (kennyadsl) [inline bin/sandbox:59]: Do we need to release a new version of solidus_i18n with the code currently in master?

**Review by kennyadsl — COMMENTED** [review]
Thanks 7️⃣ times (at least), @peterberkenbosch! I left some comments, let me know if I'm missing something.

File: Gemfile
Comment (waiting-for-dev) [reply Gemfile:16]: Yes, we must wait until new versions of `awesome_nested_set` and `solidus_i18n` are released.

**Comment by peterberkenbosch** [comment]
Thanks! Updating this PR, and will respond back to the questions, suggestions etc.

**Comment by denys-medynskyi** [comment]
hey @peterberkenbosch 
Thanks 8️⃣ times.

Do you need any help with this branch getting merged?

File: bin/sandbox
Comment (peterberkenbosch) [reply bin/sandbox:59]: I don't think so, revisiting this now. 2.1.1 should have the latest from master now anyway right?

**Comment by peterberkenbosch** [comment]
> hey @peterberkenbosch Thanks 8️⃣ times.
> 
> Do you need any help with this branch getting merged?

Not yet, rebasing all the cruft right now so we have a clean PR to work from.

File: bin/sandbox
Comment (peterberkenbosch) [reply bin/sandbox]: Confirming we don't need to do this anymore. There are a few deprecation warnings showing up, looking into those as well, but it's working as expected right now.

File: Gemfile
Comment (peterberkenbosch) [reply Gemfile:16]: `solidus_i18n` release is done btw. Still waiting on awesome_nested_set release.

**Comment by peterberkenbosch** [comment]
Should we track latest Rails 7 release by default on our main (unreleased) branch? We probably have to bump versions here too?

**Comment by waiting-for-dev** [comment]
> Should we track latest Rails 7 release by default on our main (unreleased) branch? We probably have to bump versions here too?

I'm not sure I get it. Where exactly should we bump versions?

**Comment by peterberkenbosch** [comment]
@waiting-for-dev I was thinking the Solidus version, but that's not really needed in this PR.

**Comment by peterberkenbosch** [comment]
for Rails 7.0.1 support we need a few more updates on dependencies, tracking those down now.

File: core/lib/spree/core/controller_helpers/current_host.rb
Comment (kennyadsl) [reply core/lib/spree/core/controller_helpers/current_host.rb:10]: Hey @peterberkenbosch, what was the deprecation warning?

Comment (peterberkenbosch) [reply core/lib/spree/core/controller_helpers/current_host.rb:10]: host is already set or something like that. I'm happy to remove all non-essential changes here actually.

Comment (kennyadsl) [reply core/lib/spree/core/controller_helpers/current_host.rb:10]: Just wanted to understand if that was coming from Solidus or Rails, because if it's the latter and we will receive warnings, the CI won't fail, it only fails for Solidus deprecations. cc @waiting-for-dev

**Comment by peterberkenbosch** [comment]
And just now Rails 7.0.2 is released. Getting some other errors on sandbox creation will investigate further.

```
ruby/gems/3.1.0/gems/activerecord-7.0.2/lib/active_record/database_configurations.rb:177:in `resolve_symbol_connection': The `nil` database is not configured for the `nil` environment. (ActiveRecord::AdapterNotSpecified)

  Available database configurations are:

  default
development
test
production
```

**Comment by peterberkenbosch** [comment]
Like actually just now.. https://cln.sh/EthWr9 lol

**Comment by denys-medynskyi** [comment]
@peterberkenbosch seems like you have some ruby version conflict 🤔 [here](https://app.netlify.com/sites/solidus-docs/deploys/6203987ea3cede000846fd50)
```
12:34:00 PM: Bundler found conflicting requirements for the Ruby version:
12:34:00 PM:   In Gemfile:
12:34:00 PM:     Ruby
12:34:00 PM:     solidus was resolved to 3.2.0.alpha, which depends on
12:34:00 PM:       Ruby (>= 2.7.0)
```

**Comment by peterberkenbosch** [comment]
@denys-medynskyi correct. I have added that as a todo item on this PR description. That's on Netlify side.

**Comment by jarednorman** [comment]
We should probably try to pull out the pieces of this that are ready to merge and continue tracking until the last bits are done.

**Comment by denys-medynskyi** [comment]
What's left here?

**Comment by peterberkenbosch** [comment]
Cleared up all the commits and rebased with master again. 

We have a few outstanding items still regarding the ruby version for Netlify, @jarednorman or @kennyadsl can you or someone else who has access update the build for that? We need to update Ruby 2.7 there. 

The other issue is the dummy app generation is not using the installer it seems and does not have the needed image_magick config setting.

**Comment by kennyadsl** [comment]
@peterberkenbosch Thanks! I update the netlify Ruby version for the docs site. 

> The other issue is the dummy app generation is not using the installer it seems and does not have the needed image_magick config setting.

Not sure to follow, can you please help me understand better?

**Comment by peterberkenbosch** [comment]
> @peterberkenbosch Thanks! I update the netlify Ruby version for the docs site.
> 
> > The other issue is the dummy app generation is not using the installer it seems and does not have the needed image_magick config setting.
> 
> Not sure to follow, can you please help me understand better?

Thanks @kennyadsl !

It looks like the specs are failing because it's now using the default rails 7 config for active_storage (:vips), with the Solidus installer update here I added a line that updates the config for active storage to use mini_magick instead so it's compatible with Solidus. The dummy app for the specs needs that config setting as well. I think that's the last little piece needed to have it all done.

**Comment by peterberkenbosch** [comment]
Actually, we might need to update the docker image to also contain libvips, since it's needed with Rails 7

```
LoadError:
  Could not open library 'vips.so.42': vips.so.42: cannot open shared object file: No such file or directory.
  Could not open library 'libvips.so.42': libvips.so.42: cannot open shared object file: No such file or directory
```
CircleCI also issues a warning: `You’re using a [deprecated Docker convenience image.](https://discuss.circleci.com/t/legacy-convenience-image-deprecation/41034) Upgrade to a next-gen Docker convenience image.`

File: core/lib/generators/solidus/install/install_generator.rb
Comment (waiting-for-dev) [inline core/lib/generators/solidus/install/install_generator.rb:59]: Hey, @peterberkenbosch, that could cause a conflict with the user's preference, right? We should support both by checking the configured value

**Comment by waiting-for-dev** [comment]
Hey @peterberkenbosch, I added a message about the `vips` vs. `mini_magick` issue. I have time to work on that now. Would you mind if I add commits to your PR?

**Comment by peterberkenbosch** [comment]
@waiting-for-dev we could update the `style_to_size` private method in `core/app/models/concerns/spree/active_storage_adapter/attachment.rb` to check if the `Rails.application.config.active_storage.variant_processor` == ':vips` to make sure we return integers there and not a String.

**Comment by peterberkenbosch** [comment]
also headsup @waiting-for-dev I rebased master here. (Will wait with that the next time, force of habbit).

**Comment by waiting-for-dev** [comment]
> @waiting-for-dev we could update the `style_to_size` private method in `core/app/models/concerns/spree/active_storage_adapter/attachment.rb` to check if the `Rails.application.config.active_storage.variant_processor` == ':vips` to make sure we return integers there and not a String.

Cool, I saw the error, but I hadn't had the time to dig into it. I'll do it tomorrow :slightly_smiling_face:

**Comment by waiting-for-dev** [comment]
Task linked: [CU-24c99ek Rails 7 support](https://app.clickup.com/t/24c99ek)

**Comment by waiting-for-dev** [comment]
@peterberkenbosch, so rails 6.1 already introduced vips as an adapter, but we're noticing the issue here because it's the default on Rails 7.0. We need to fix the underlying issue and rebase from there. Please, see https://github.com/solidusio/solidus/pull/4317 (and related https://github.com/solidusio/solidus/pull/4318)

**Comment by waiting-for-dev** [comment]
Hey @peterberkenbosch, now my turn, I forced a push to be able to incorporate the fix for vips :blush: It seems there are other errors. I'm going to look into them.

**Comment by peterberkenbosch** [comment]
Thanks @waiting-for-dev I'm reading along :)

**Comment by waiting-for-dev** [comment]
Hey @peterberkenbosch, everything should be ready now :crossed_fingers: I forced a push again to restructure the commits to be more easily reviewed.

@solidusio/core-team I'm also looking forward to reading your feedback. Please, proceed commit by commit to understand the changes' cohesion better.

**Comment by peterberkenbosch** [comment]
Awesome @waiting-for-dev 🥳

**Comment by waiting-for-dev** [comment]
Not caught by the test suite, but we still need to make deface compatible with Rails 7. Try to launch the sandbox app to reproduce. See https://github.com/spree/deface/issues/225.

**Comment by waiting-for-dev** [comment]
> Not caught by the test suite, but we still need to make deface compatible with Rails 7. Try to launch the sandbox app to reproduce. See [spree/deface#225](https://github.com/spree/deface/issues/225).

That has already been fixed, thanks to the fantastic work of @elia! Technically, deface is not a dependency of solidus, but many extensions are using it. One example is `solidus_auth_devise`, which is used in the sandbox application. That explains why the test suite was green, but the sandbox didn't work. Now, it's working again :heart:

**Comment by denys-medynskyi** [comment]
Looks 💚 does it mean it is close to release?

**Comment by peterberkenbosch** [comment]
Looks like it!

**Comment by peterberkenbosch** [comment]
@waiting-for-dev just rebased master branch and updated the merge conflicts.

**Comment by peterberkenbosch** [comment]
@waiting-for-dev I think this is ready for final round review / merge?

**Comment by denys-medynskyi** [comment]
Looks like failed test is a false positive 🤔 pr maybe `eql` shouldn't be used in this test, because as I see it is just different instance...
```
Failure/Error: expect(shipment.stock_location).not_to eql stock_item.stock_location

  expected: value != #<Spree::StockLocation id: 1998, name: "NY Warehouse", created_at: "2022-04-08 10:41:01", updated_at:..., position: 1, restock_inventory: true, fulfillable: true, code: nil, check_stock_on_transfer: true>
       got: #<Spree::StockLocation id: 1998, name: "NY Warehouse", created_at: "2022-04-08 10:41:01", updated_at:..., position: 1, restock_inventory: true, fulfillable: true, code: nil, check_stock_on_transfer: true>

  (compared using eql?)

  Diff:
  @@ -1,4 +1,4 @@
  -#<Spree::StockLocation:0x0000564a1895da40
  +#<Spree::StockLocation:0x0000564a18960ad8
    id: 1998,
    name: "NY Warehouse",
    created_at: Fri, 08 Apr 2022 10:41:01 UTC +00:00,
./spec/models/spree/order_inventory_spec.rb:181:in `block (6 levels) in <top (required)>'
```

**Comment by kennyadsl** [comment]
@denys-medynskyi should be a flaky, re-running!

**Comment by denys-medynskyi** [comment]
It is green now 💚

**Comment by denys-medynskyi** [comment]
@kennyadsl can you merge it please 🙏

**Review by kennyadsl — APPROVED** [review]
👍 Approved.

@denys-medynskyi we need to wait for a review from another core team member before merging.

**Comment by waiting-for-dev** [comment]
I added a new commit reverting dropping support for Ruby 2.5 & Ruby 2.6. We still support Rails 5.2, which works from 2.5 onwards. We'll deprecate Ruby & Rails versions on major releases, as that's backward-incompatible changes.

**Review by jarednorman — APPROVED** [review]

**Comment by denys-medynskyi** [comment]
We are good to merge 🎉

**Comment by mslinn** [comment]
When might this merged PR get packaged into a released gem? Solidus 3.1.7 still uses Rail 6.1.

**Comment by waiting-for-dev** [comment]
> When might this merged PR get packaged into a released gem? Solidus 3.1.7 still uses Rail 6.1.

@mslinn, we're wrapping everything up and you can expect a new release in the next couple of weeks :crossed_fingers:

**Comment by jarednorman** [comment]
You can point your Gemfile at the latest code from GitHub, yes. Some stores do that to stay as up to date as possible.

**Comment by mslinn** [comment]
I wrote up a [detailed explanation](https://mslinn.com/blog/2022/07/22/solidus.html#addSolidus) of what I did to build the gems, and created a [GitHub discussion](https://github.com/solidusio/solidus/discussions/4464) to figure out why `solidus_core` is not loaded when added to a Rails app, but did not get much response. Might someone take a moment to suggest ideas? ... Thanks!

**Comment by asfour75** [comment]
> > When might this merged PR get packaged into a released gem? Solidus 3.1.7 still uses Rail 6.1.
> 
> @mslinn, we're wrapping everything up and you can expect a new release in the next couple of weeks crossed_fingers

Thank you all for the great effort pushing solidus to suppoer rails 7 as soon as possible .. 
It's been almost a year .. Yet I'm still searching for offical announcement .. are we there yet ?

**Comment by kennyadsl** [comment]
Rails 7 support had been merged and release a while ago, maybe in 3.2 if I recall correctly.

**Comment by BenMorganIO** [comment]
@asfour75 I think we're on Rails 7 with Solidus 3.3 but we ran 3.2 with 7 as well.

**Comment by asfour75** [comment]
> Rails 7 support had been merged and released a while ago, maybe in 3.2 if I recall correctly.

Based on the offical gihub page https://github.com/solidusio/solidus
it says:

To add Solidus, begin with a Rails 5.2, 6 or 6.1 application and a database configured and created.

so I assumed it only supports up to 6.1
Nevertheless .. I tried it yesterday on rails 7 and it worekd perfectly .. 
Thanks for the note .. nice work

**Comment by waiting-for-dev** [comment]
Thanks for reporting. The README is currently being fixed at https://github.com/solidusio/solidus/pull/5012/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5

**Pull request merged**
